jni/src/faiss_wrapper.cpp

```diff
@@ -108,63 +110,61 @@ void knn_jni::faiss_wrapper::CreateIndex(knn_jni::JNIUtilInterface * jniUtil, JN
+    // Thread count
+    int threadCount = 0;
     if(parametersCpp.find(knn_jni::INDEX_THREAD_QUANTITY) != parametersCpp.end()) {
-        auto threadCount = jniUtil->ConvertJavaObjectToCppInteger(env, parametersCpp[knn_jni::INDEX_THREAD_QUANTITY]);
-        omp_set_num_threads(threadCount);
+        threadCount = jniUtil->ConvertJavaObjectToCppInteger(env, parametersCpp[knn_jni::INDEX_THREAD_QUANTITY]);
     }
 
-    // Add extra parameters that cant be configured with the index factory
-    if(parametersCpp.find(knn_jni::PARAMETERS) != parametersCpp.end()) {
-        jobject subParametersJ = parametersCpp[knn_jni::PARAMETERS];
-        auto subParametersCpp = jniUtil->ConvertJavaMapToCppMap(env, subParametersJ);
-        SetExtraParameters(jniUtil, env, subParametersCpp, indexWriter.get());
-        jniUtil->DeleteLocalRef(env, subParametersJ);
-    }
-    jniUtil->DeleteLocalRef(env, parametersJ);
+    // Vectors address
+    int64_t vectorsAddress = (int64_t)vectorsAddressJ;
 
-    // Check that the index does not need to be trained
-    if(!indexWriter->is_trained) {
-        throw std::runtime_error("Index is not trained");
+    // Ids
+    auto ids = jniUtil->ConvertJavaIntArrayToCppIntVector(env, idsJ);
+
+    // Index path
+    std::string indexPathCpp(jniUtil->ConvertJavaStringToCppString(env, indexPathJ));
+
+    // Extra parameters
+    // TODO: parse the entire map and remove jni object
+    std::unordered_map<std::string, jobject> subParametersCpp;
+    if(parametersCpp.find(knn_jni::PARAMETERS) != parametersCpp.end()) {
+        subParametersCpp = jniUtil->ConvertJavaMapToCppMap(env, parametersCpp[knn_jni::PARAMETERS]);
     }
+    // end parameters to pass
 
-    auto idVector = jniUtil->ConvertJavaIntArrayToCppIntVector(env, idsJ);
-    faiss::IndexIDMap idMap = faiss::IndexIDMap(indexWriter.get());
-    idMap.add_with_ids(numVectors, inputVectors->data(), idVector.data());
+    // Create IndexCreateService based on the index description
+    bool isBinaryIndex = indexDescriptionCpp[0] == 'B';
+    std::unique_ptr<IndexService> indexService = indexServiceFactory->createIndexService(isBinaryIndex);
+    indexService->createIndex(jniUtil, env, metric, indexDescriptionCpp, dim, numIds, threadCount, vectorsAddress, ids, indexPathCpp, subParametersCpp);
 
-    // Write the index to disk
-    std::string indexPathCpp(jniUtil->ConvertJavaStringToCppString(env, indexPathJ));
-    faiss::write_index(&idMap, indexPathCpp.c_str());
     // Releasing the vectorsAddressJ memory as that is not required once we have created the index.
     // This is not the ideal approach, please refer this gh issue for long term solution:
     // https://github.com/opensearch-project/k-NN/issues/1600
-    delete inputVectors;
+    if (isBinaryIndex) {
+        delete reinterpret_cast<std::vector<uint8_t>*>(vectorsAddress);
+    }
+    else {
+        delete reinterpret_cast<std::vector<float>*>(vectorsAddress);
+    }
```

Comment:
why we need to `reinterpret_cast` the casting here? We can just call delete and then the address?